src/main/java/za/co/neslotech/spring/trainspringbootinsight/controller/AGenericController.java

```diff
@@ -0,0 +1,69 @@
+package za.co.neslotech.spring.trainspringbootinsight.controller;
+
+import jakarta.persistence.EntityNotFoundException;
+import org.springframework.beans.BeanUtils;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+import java.util.List;
+
+/**
+ * A generic REST API Controller with CRUD operations.
+ * The child class should use the following annotations
+ *      1. @RestController
+ * @param <T> The entity class.
+ * @param <I> The primary key field type.
+ */
+public abstract class AGenericController <T, I> {
+
+    protected abstract JpaRepository<T, I> getRepository();
+    protected abstract String getEntityNotFoundDescription(final I id);
+    protected abstract String[] getUpdateColumnExclusions();
+
+    @GetMapping
+    public List<T> list(){
```

Comment:
We generally follow the pattern of using returning ResponseEntity objects within Spring Boot controllers

Returning a ResponseEntity object in Spring Boot RESTful API controllers can provide more control and flexibility over the response that gets sent back to the client.

A ResponseEntity allows you to set the HTTP status code, headers, and body of the response explicitly. This means that you can customize the response to include additional metadata or error messages, and also return the appropriate status code based on the success or failure of the request.

For example, if a resource is not found, you can return a ResponseEntity with a 404 Not Found status code and a custom message in the body. On the other hand, if a resource is successfully created, you can return a ResponseEntity with a 201 Created status code and a location header that points to the newly created resource.

In addition, returning a ResponseEntity object also allows you to support content negotiation by specifying the media type of the response body. This enables clients to request different representations of the same resource, such as JSON or XML.

Overall, returning a ResponseEntity object provides a more flexible and robust way of handling HTTP responses in Spring Boot RESTful APIs.